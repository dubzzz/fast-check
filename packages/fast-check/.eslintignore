bundle/
node_modules/
lib/
lib-*/
perf/
jest*.js
jest*.cjs