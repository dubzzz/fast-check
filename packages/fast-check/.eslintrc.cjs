module.exports = {
  extends: [
    'eslint:recommended',
    'plugin:@typescript-eslint/eslint-recommended',
    'plugin:@typescript-eslint/recommended',
    'prettier',
  ],
  rules: {
    '@typescript-eslint/explicit-module-boundary-types': 'error',
    '@typescript-eslint/explicit-function-return-type': 'off',
    '@typescript-eslint/interface-name-prefix': 'off',
    '@typescript-eslint/no-empty-function': 'warn',
    '@typescript-eslint/no-empty-interface': 'off',
    '@typescript-eslint/no-explicit-any': 'off',
    '@typescript-eslint/no-this-alias': 'warn',
    '@typescript-eslint/no-unused-vars': ['error', { argsIgnorePattern: '^_' }],
    '@typescript-eslint/no-use-before-define': 'warn',
  },
  overrides: [
    {
      files: ['*.spec.ts', 'test/unit/**/*.ts', 'test/e2e/**/*.ts'],
      rules: {
        '@typescript-eslint/no-empty-function': 'off',
        '@typescript-eslint/no-use-before-define': 'off',
        '@typescript-eslint/no-non-null-assertion': 'off',
      },
    },
  ],
};
