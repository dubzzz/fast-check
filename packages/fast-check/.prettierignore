bundle/
coverage/
docs/
documentation/Arbitraries.md
lib/
lib-*/