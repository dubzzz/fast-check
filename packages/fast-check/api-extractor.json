{
  "$schema": "https://developer.microsoft.com/json-schemas/api-extractor/v7/api-extractor.schema.json",
  "compiler": {
    "tsconfigFilePath": "<projectFolder>/tsconfig.publish.types.json"
  },
  "tsdocMetadata": {
    "enabled": false
  },
  "mainEntryPointFilePath": "lib/types/fast-check-default.d.ts",
  "dtsRollup": {
    "enabled": false
  },
  "docModel": {
    "enabled": true,
    "apiJsonFilePath": "docs/fast-check.api.json"
  },
  "apiReport": {
    "enabled": false
  },
  "messages": {
    "compilerMessageReporting": {
      "default": {
        "logLevel": "error"
      }
    },
    "extractorMessageReporting": {
      "default": {
        "logLevel": "error"
      },
      "ae-unresolved-link": {
        "logLevel": "warning"
      }
    },
    "tsdocMessageReporting": {
      "default": {
        "logLevel": "error"
      }
    }
  }
}
