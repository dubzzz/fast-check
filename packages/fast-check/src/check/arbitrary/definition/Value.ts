import type { WithCloneMethod } from '../../symbols';
import { cloneMethod, hasCloneMethod } from '../../symbols';

const safeObjectDefineProperty = Object.defineProperty;

/**
 * A `Value<T, TShrink = T>` holds an internal value of type `T`
 * and its associated context
 *
 * @remarks Since 3.0.0 (previously called `NextValue` in 2.15.0)
 * @public
 */
export class Value<T> {
  /**
   * State storing the result of hasCloneMethod
   * If `true` the value will be cloned each time it gets accessed
   * @remarks Since 2.15.0
   */
  readonly hasToBeCloned: boolean;
  /**
   * Flag indicating whether or not the this.value has already been called once
   * If so, the underlying will be cloned
   * Only set when hasToBeCloned = true
   */
  private readOnce: boolean;
  /**
   * Safe value of the shrinkable
   * Depending on `hasToBeCloned` it will either be `value_` or a clone of it
   * @remarks Since 2.15.0
   */
  readonly value!: T;
  /**
   * Internal value of the shrinkable
   * @remarks Since 2.15.0
   */
  readonly value_: T;
  /**
   * Context for the generated value
   * TODO - Do we want to clone it too?
   * @remarks 2.15.0
   */
  readonly context: unknown;

  /**
   * @param value_ - Internal value of the shrinkable
   * @param context - Context associated to the generated value (useful for shrink)
   * @param customGetValue - Limited to internal usages (to ease migration to next), it will be removed on next major
   */
  constructor(value_: T, context: unknown, customGetValue: (() => T) | undefined = undefined) {
    this.value_ = value_;
    this.context = context;
    this.hasToBeCloned = customGetValue !== undefined || hasCloneMethod(value_);
    this.readOnce = false;
    if (this.hasToBeCloned) {
      safeObjectDefineProperty(this, 'value', { get: customGetValue !== undefined ? customGetValue : this.getValue });
    } else {
      this.value = value_;
    }
  }

  /** @internal */
  private getValue() {
    if (this.hasToBeCloned) {
      if (!this.readOnce) {
        this.readOnce = true;
        return this.value_;
      }
      return (this.value_ as unknown as WithCloneMethod<T>)[cloneMethod]();
    }
    return this.value_;
  }
}
