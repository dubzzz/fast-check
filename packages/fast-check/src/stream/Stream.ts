import {
  filterHelper,
  flatMapHelper,
  joinHelper,
  mapHelper,
  nilHelper,
  takeNHelper,
  takeWhileHelper,
} from './StreamHelpers';

const safeSymbolIterator: typeof Symbol.iterator = Symbol.iterator;

/**
 * Wrapper around `IterableIterator` interface
 * offering a set of helpers to deal with iterations in a simple way
 *
 * @remarks Since 0.0.7
 * @public
 */
export class Stream<T> implements IterableIterator<T> {
  /**
   * Create an empty stream of T
   * @remarks Since 0.0.1
   */
  static nil<T>(): Stream<T> {
    return new Stream(nilHelper());
  }

  /**
   * Create a stream of T from a variable number of elements
   *
   * @param elements - Elements used to create the Stream
   * @remarks Since 2.12.0
   */
  static of<T>(...elements: T[]): Stream<T> {
    return new Stream(elements[safeSymbolIterator]());
  }

  // /*DEBUG*/ // no double iteration
  // /*DEBUG*/ private isLive: boolean;

  /**
   * Create a Stream based on `g`
   * @param g - Underlying data of the Stream
   */
  constructor(/** @internal */ private readonly g: IterableIterator<T>) {
    // /*DEBUG*/ this.isLive = true;
  }

  // /*DEBUG*/ private closeCurrentStream() {
  // /*DEBUG*/   if (! this.isLive) throw new Error('Stream has already been closed');
  // /*DEBUG*/   this.isLive = false;
  // /*DEBUG*/ }

  next(): IteratorResult<T> {
    return this.g.next();
  }
  [safeSymbolIterator](): IterableIterator<T> {
    // /*DEBUG*/ this.closeCurrentStream();
    return this.g;
  }

  /**
   * Map all elements of the Stream using `f`
   *
   * WARNING: It closes the current stream
   *
   * @param f - Mapper function
   * @remarks Since 0.0.1
   */
  map<U>(f: (v: T) => U): Stream<U> {
    // /*DEBUG*/ this.closeCurrentStream();
    return new Stream(mapHelper(this.g, f));
  }
  /**
   * Flat map all elements of the Stream using `f`
   *
   * WARNING: It closes the current stream
   *
   * @param f - Mapper function
   * @remarks Since 0.0.1
   */
  flatMap<U>(f: (v: T) => IterableIterator<U>): Stream<U> {
    // /*DEBUG*/ this.closeCurrentStream();
    return new Stream(flatMapHelper(this.g, f));
  }

  /**
   * Drop elements from the Stream while `f(element) === true`
   *
   * WARNING: It closes the current stream
   *
   * @param f - Drop condition
   * @remarks Since 0.0.1
   */
  dropWhile(f: (v: T) => boolean): Stream<T> {
    let foundEligible = false;
    function* helper(v: T): IterableIterator<T> {
      if (foundEligible || !f(v)) {
        foundEligible = true;
        yield v;
      }
    }
    return this.flatMap(helper);
  }
  /**
   * Drop `n` first elements of the Stream
   *
   * WARNING: It closes the current stream
   *
   * @param n - Number of elements to drop
   * @remarks Since 0.0.1
   */
  drop(n: number): Stream<T> {
    if (n <= 0) {
      return this;
    }
    let idx = 0;
    function helper(): boolean {
      return idx++ < n;
    }
    return this.dropWhile(helper);
  }
  /**
   * Take elements from the Stream while `f(element) === true`
   *
   * WARNING: It closes the current stream
   *
   * @param f - Take condition
   * @remarks Since 0.0.1
   */
  takeWhile(f: (v: T) => boolean): Stream<T> {
    // /*DEBUG*/ this.closeCurrentStream();
    return new Stream(takeWhileHelper(this.g, f));
  }
  /**
   * Take `n` first elements of the Stream
   *
   * WARNING: It closes the current stream
   *
   * @param n - Number of elements to take
   * @remarks Since 0.0.1
   */
  take(n: number): Stream<T> {
    // /*DEBUG*/ this.closeCurrentStream();
    return new Stream(takeNHelper(this.g, n));
  }

  /**
   * Filter elements of the Stream
   *
   * WARNING: It closes the current stream
   *
   * @param f - Elements to keep
   * @remarks Since 1.23.0
   */
  filter<U extends T>(f: (v: T) => v is U): Stream<U>;
  /**
   * Filter elements of the Stream
   *
   * WARNING: It closes the current stream
   *
   * @param f - Elements to keep
   * @remarks Since 0.0.1
   */
  filter(f: (v: T) => boolean): Stream<T>;
  filter<U extends T>(f: (v: T) => v is U): Stream<U> {
    // /*DEBUG*/ this.closeCurrentStream();
    return new Stream(filterHelper(this.g, f));
  }

  /**
   * Check whether all elements of the Stream are successful for `f`
   *
   * WARNING: It closes the current stream
   *
   * @param f - Condition to check
   * @remarks Since 0.0.1
   */
  every(f: (v: T) => boolean): boolean {
    // /*DEBUG*/ this.closeCurrentStream();
    for (const v of this.g) {
      if (!f(v)) {
        return false;
      }
    }
    return true;
  }
  /**
   * Check whether one of the elements of the Stream is successful for `f`
   *
   * WARNING: It closes the current stream
   *
   * @param f - Condition to check
   * @remarks Since 0.0.1
   */
  has(f: (v: T) => boolean): [boolean, T | null] {
    // /*DEBUG*/ this.closeCurrentStream();
    for (const v of this.g) {
      if (f(v)) {
        return [true, v];
      }
    }
    return [false, null];
  }

  /**
   * Join `others` Stream to the current Stream
   *
   * WARNING: It closes the current stream and the other ones (as soon as it iterates over them)
   *
   * @param others - Streams to join to the current Stream
   * @remarks Since 0.0.1
   */
  join(...others: IterableIterator<T>[]): Stream<T> {
    // /*DEBUG*/ this.closeCurrentStream();
    return new Stream(joinHelper(this.g, others));
  }

  /**
   * Take the `nth` element of the Stream of the last (if it does not exist)
   *
   * WARNING: It closes the current stream
   *
   * @param nth - Position of the element to extract
   * @remarks Since 0.0.12
   */
  getNthOrLast(nth: number): T | null {
    // /*DEBUG*/ this.closeCurrentStream();
    let remaining = nth;
    let last: T | null = null;
    for (const v of this.g) {
      if (remaining-- === 0) return v;
      last = v;
    }
    return last;
  }
}

/**
 * Create a Stream based on `g`
 *
 * @param g - Underlying data of the Stream
 *
 * @remarks Since 0.0.7
 * @public
 */
export function stream<T>(g: IterableIterator<T>): Stream<T> {
  return new Stream<T>(g);
}
