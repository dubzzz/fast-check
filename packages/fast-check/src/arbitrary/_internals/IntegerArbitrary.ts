import { Arbitrary } from '../../check/arbitrary/definition/Arbitrary';
import { Value } from '../../check/arbitrary/definition/Value';
import type { Random } from '../../random/generator/Random';
import { Stream } from '../../stream/Stream';
import { integerLogLike, biasNumericRange } from './helpers/BiasNumericRange';
import { shrinkInteger } from './helpers/ShrinkInteger';

const safeMathSign = Math.sign;
const safeNumberIsInteger = Number.isInteger;
const safeObjectIs = Object.is;

/** @internal */
export class IntegerArbitrary extends Arbitrary<number> {
  constructor(
    readonly min: number,
    readonly max: number,
  ) {
    super();
  }

  generate(mrng: Random, biasFactor: number | undefined): Value<number> {
    const range = this.computeGenerateRange(mrng, biasFactor);
    return new Value(mrng.nextInt(range.min, range.max), undefined);
  }

  canShrinkWithoutContext(value: unknown): value is number {
    return (
      typeof value === 'number' &&
      safeNumberIsInteger(value) &&
      !safeObjectIs(value, -0) &&
      this.min <= value &&
      value <= this.max
    );
  }

  shrink(current: number, context?: unknown): Stream<Value<number>> {
    if (!IntegerArbitrary.isValidContext(current, context)) {
      // No context:
      //   Take default target and shrink towards it
      //   Try the target on first try
      const target = this.defaultTarget();
      return shrinkInteger(current, target, true);
    }
    if (this.isLastChanceTry(current, context)) {
      // Last chance try...
      // context is set to undefined, so that shrink will restart
      // without any assumptions in case our try find yet another bug
      return Stream.of(new Value(context, undefined));
    }
    // Normal shrink process
    return shrinkInteger(current, context, false);
  }

  private defaultTarget(): number {
    // min <= 0 && max >= 0   => shrink towards zero
    if (this.min <= 0 && this.max >= 0) {
      return 0;
    }
    // min < 0                => shrink towards max (closer to zero)
    // otherwise              => shrink towards min (closer to zero)
    return this.min < 0 ? this.max : this.min;
  }

  private computeGenerateRange(mrng: Random, biasFactor: number | undefined): { min: number; max: number } {
    if (biasFactor === undefined || mrng.nextInt(1, biasFactor) !== 1) {
      return { min: this.min, max: this.max };
    }
    const ranges = biasNumericRange(this.min, this.max, integerLogLike);
    if (ranges.length === 1) {
      return ranges[0];
    }
    const id = mrng.nextInt(-2 * (ranges.length - 1), ranges.length - 2); // 1st range has the highest priority
    return id < 0 ? ranges[0] : ranges[id + 1];
  }

  private isLastChanceTry(current: number, context: number): boolean {
    // If true...
    // We already reached what we thought to be the minimal failing value.
    // But in-between other values may have shrunk (values coming from other arbitraries).
    // In order to check if they impacted us, we just try to move very close to our current value.
    // It is not ideal but it can help restart a shrinking process that stopped too early.
    if (current > 0) return current === context + 1 && current > this.min;
    if (current < 0) return current === context - 1 && current < this.max;
    return false;
  }

  private static isValidContext(current: number, context?: unknown): context is number {
    // Context contains a value between zero and current that is known to be
    // the closer to zero passing value*.
    // *More precisely: our shrinker will not try something closer to zero
    if (context === undefined) {
      return false;
    }
    if (typeof context !== 'number') {
      throw new Error(`Invalid context type passed to IntegerArbitrary (#1)`);
    }
    if (context !== 0 && safeMathSign(current) !== safeMathSign(context)) {
      throw new Error(`Invalid context value passed to IntegerArbitrary (#2)`);
    }
    return true;
  }
}
