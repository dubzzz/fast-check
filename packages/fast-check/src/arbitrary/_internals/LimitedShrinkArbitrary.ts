import { Arbitrary } from '../../check/arbitrary/definition/Arbitrary';
import { Value } from '../../check/arbitrary/definition/Value';
import type { Random } from '../../random/generator/Random';
import { Stream } from '../../stream/Stream';
import { zipIterableIterators } from './helpers/ZipIterableIterators';

/** @internal */
function* iotaFrom(startValue: number) {
  let value = startValue;
  while (true) {
    yield value;
    ++value;
  }
}

/** @internal */
export class LimitedShrinkArbitrary<T> extends Arbitrary<T> {
  constructor(
    readonly arb: Arbitrary<T>,
    readonly maxShrinks: number,
  ) {
    super();
  }
  generate(mrng: Random, biasFactor: number | undefined): Value<T> {
    const value = this.arb.generate(mrng, biasFactor);
    return this.valueMapper(value, 0);
  }
  canShrinkWithoutContext(value: unknown): value is T {
    return this.arb.canShrinkWithoutContext(value);
  }
  shrink(value: T, context?: unknown): Stream<Value<T>> {
    if (this.isSafeContext(context)) {
      return this.safeShrink(value, context.originalContext, context.length);
    }
    return this.safeShrink(value, undefined, 0);
  }
  private safeShrink(value: T, originalContext: unknown, currentLength: number): Stream<Value<T>> {
    const remaining = this.maxShrinks - currentLength;
    if (remaining <= 0) {
      return Stream.nil(); // early-exit to avoid potentially expensive computations in .shrink
    }
    return new Stream(zipIterableIterators(this.arb.shrink(value, originalContext), iotaFrom(currentLength + 1)))
      .take(remaining)
      .map((valueAndLength) => this.valueMapper(valueAndLength[0], valueAndLength[1]));
  }
  private valueMapper(v: Value<T>, newLength: number): Value<T> {
    const context: LimitedShrinkArbitraryContext = { originalContext: v.context, length: newLength };
    return new Value(v.value, context);
  }
  private isSafeContext(context: unknown): context is LimitedShrinkArbitraryContext {
    return (
      context != null &&
      typeof context === 'object' &&
      'originalContext' in (context as any) &&
      'length' in (context as any)
    );
  }
}

/** @internal */
type LimitedShrinkArbitraryContext = {
  originalContext: unknown;
  length: number;
};
