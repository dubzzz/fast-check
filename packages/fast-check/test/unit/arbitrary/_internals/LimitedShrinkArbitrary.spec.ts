import { describe, it, expect } from 'vitest';
import * as fc from 'fast-check';
import { Value } from '../../../../src/check/arbitrary/definition/Value';
import { Stream } from '../../../../src/stream/Stream';
import { fakeRandom } from '../__test-helpers__/RandomHelpers';

import { declareCleaningHooksForSpies } from '../__test-helpers__/SpyCleaner';
import { fakeArbitrary } from '../__test-helpers__/ArbitraryHelpers';
import { LimitedShrinkArbitrary } from '../../../../src/arbitrary/_internals/LimitedShrinkArbitrary';
import { IntegerArbitrary } from '../../../../src/arbitrary/_internals/IntegerArbitrary';
import {
  assertProduceSameValueGivenSameSeed,
  assertShrinkProducesSameValueWithoutInitialContext,
} from '../__test-helpers__/ArbitraryAssertions';
import { buildShrinkTree, renderTree } from '../__test-helpers__/ShrinkTree';

describe('LimitedShrinkArbitrary', () => {
  declareCleaningHooksForSpies();

  describe('generate', () => {
    it('should only rely on the underlying arbitrary to generate values and forward the target value as-is', () => {
      fc.assert(
        fc.property(
          fc.anything(),
          fc.anything(),
          fc.option(fc.integer({ min: 2 }), { nil: undefined }),
          fc.nat(),
          (generatedValue, context, biasFactor, maxShrinks) => {
            // Arrange
            const value = new Value(generatedValue, context);
            const { instance: mrng } = fakeRandom();
            const { instance: arbitrary, generate } = fakeArbitrary();
            generate.mockReturnValueOnce(value);

            // Act
            const arb = new LimitedShrinkArbitrary(arbitrary, maxShrinks);
            const out = arb.generate(mrng, biasFactor);

            // Assert
            expect(generate).toHaveBeenCalledTimes(1);
            expect(generate).toHaveBeenCalledWith(mrng, biasFactor);
            expect(out.value_).toBe(generatedValue); // Remark: The instance of Value itself might be different
          },
        ),
      );
    });
  });

  describe('canShrinkWithoutContext', () => {
    it('should only rely on the underlying arbitrary to check if it can generate a value without any context', () => {
      fc.assert(
        fc.property(fc.anything(), fc.boolean(), fc.nat(), (assessedValue, expectedOutput, maxShrinksPerLevel) => {
          // Arrange
          const { instance: arbitrary, canShrinkWithoutContext } = fakeArbitrary();
          canShrinkWithoutContext.mockReturnValueOnce(expectedOutput);

          // Act
          const arb = new LimitedShrinkArbitrary(arbitrary, maxShrinksPerLevel);
          const out = arb.canShrinkWithoutContext(assessedValue);

          // Assert
          expect(canShrinkWithoutContext).toHaveBeenCalledTimes(1);
          expect(canShrinkWithoutContext).toHaveBeenCalledWith(assessedValue);
          expect(out).toBe(expectedOutput);
        }),
      );
    });
  });

  describe('shrink', () => {
    it('should call shrink of the underlying arbitrary but cut it to maxShrinks when >=1 and receiving its own values of first level', () => {
      fc.assert(
        fc.property(
          fc.anything(),
          fc.anything(),
          fc.option(fc.integer({ min: 2 }), { nil: undefined }),
          fc.integer({ min: 1 }),
          fc.array(fc.tuple(fc.anything(), fc.anything())),
          (generatedValue, context, biasFactor, maxShrinks, shrunkValues) => {
            // Arrange
            const value = new Value(generatedValue, context);
            const { instance: mrng } = fakeRandom();
            const { instance: arbitrary, generate, shrink } = fakeArbitrary();
            generate.mockReturnValueOnce(value);
            shrink.mockReturnValueOnce(Stream.of(...shrunkValues.map(([value, context]) => new Value(value, context))));

            // Act
            const arb = new LimitedShrinkArbitrary(arbitrary, maxShrinks);
            const g = arb.generate(mrng, biasFactor);
            const shrinks = arb.shrink(g.value_, g.context);

            // Assert
            expect(shrink).toHaveBeenCalledTimes(1);
            expect(shrink).toHaveBeenCalledWith(generatedValue, context);
            expect([...shrinks].map((s) => s.value_)).toEqual(shrunkValues.map((s) => s[0]).slice(0, maxShrinks)); // Remark: The instance of Value itself might be different
          },
        ),
      );
    });

    it('should not call shrink of the underlying arbitrary when no shrinks even if receiving its own values', () => {
      fc.assert(
        fc.property(
          fc.anything(),
          fc.anything(),
          fc.option(fc.integer({ min: 2 }), { nil: undefined }),
          (generatedValue, context, biasFactor) => {
            // Arrange
            const value = new Value(generatedValue, context);
            const { instance: mrng } = fakeRandom();
            const { instance: arbitrary, generate, shrink } = fakeArbitrary();
            generate.mockReturnValueOnce(value);

            // Act
            const arb = new LimitedShrinkArbitrary(arbitrary, 0);
            const g = arb.generate(mrng, biasFactor);
            const shrinks = arb.shrink(g.value_, g.context);

            // Assert
            expect(shrink).not.toHaveBeenCalled();
            expect([...shrinks].map((s) => s.value_)).toHaveLength(0);
          },
        ),
      );
    });

    it('should cut the shrinker to only produce at most maxShrinks shrinks whatever the shrinking path', () => {
      fc.assert(
        fc.property(
          fc.anything(),
          fc.anything(),
          fc.option(fc.integer({ min: 2 }), { nil: undefined }),
          fc.nat({ max: 50 }), // limiting shrink count to 50 to avoid taking hours
          fc.infiniteStream(fc.infiniteStream(fc.tuple(fc.anything(), fc.anything()))),
          fc.infiniteStream(fc.integer({ min: 1 })),
          (generatedValue, context, biasFactor, maxShrinks, shrunkValuesLevels, shrinkingPath) => {
            // Arrange
            const value = new Value(generatedValue, context);
            const { instance: mrng } = fakeRandom();
            const { instance: arbitrary, generate, shrink } = fakeArbitrary();
            generate.mockReturnValueOnce(value);
            shrink.mockImplementation((): Stream<Value<any>> => {
              const itShrinks = shrunkValuesLevels.next();
              if (itShrinks.done) {
                throw new Error('Not possible, only added for typings');
              }
              return new Stream(itShrinks.value.map(([value, context]) => new Value(value, context)));
            });

            // Act / Assert
            const arb = new LimitedShrinkArbitrary(arbitrary, maxShrinks);
            const g = arb.generate(mrng, biasFactor);
            let done = false;
            let shrinkCount = 0;
            let lastValue = g;
            while (!done) {
              // Shrinking the value until we reach an end
              const shrinkLength: number = shrinkingPath.next().value;
              const shrinks = arb.shrink(lastValue.value_, lastValue.context);
              for (let i = 0; i !== shrinkLength; ++i) {
                const nextShrink = shrinks.next();
                if (nextShrink.done) {
                  done = true;
                  break;
                }
                ++shrinkCount;
                lastValue = nextShrink.value;
              }
            }
            expect(shrinkCount).toBe(maxShrinks); // exact as our underlying arbitrary is shrinking forever
          },
        ),
      );
    });
  });
});

describe('LimitedShrinkArbitrary (integration)', () => {
  type Extra = { maxShrinks: number };
  const extraParameters: fc.Arbitrary<Extra> = fc.record({ maxShrinks: fc.nat() });

  const limitedShrinkBuilder = (extra: Extra) =>
    new LimitedShrinkArbitrary(new IntegerArbitrary(0, 0x7fffffff), extra.maxShrinks);

  it('should produce the same values given the same seed', () => {
    assertProduceSameValueGivenSameSeed(limitedShrinkBuilder, { extraParameters });
  });

  it('should be able to shrink to the same values without initial context', () => {
    assertShrinkProducesSameValueWithoutInitialContext(limitedShrinkBuilder, { extraParameters });
  });

  it.each`
    rawValue | maxShrinks
    ${251}   | ${0}
    ${251}   | ${1}
    ${251}   | ${2}
    ${251}   | ${3}
    ${251}   | ${4}
  `('should be able to shrink $rawValue given constraints maxShrinks:$maxShrinks', ({ rawValue, maxShrinks }) => {
    // Arrange
    const arb = new LimitedShrinkArbitrary(new IntegerArbitrary(0, 0x7fffffff), maxShrinks);
    const value = new Value(rawValue, undefined);

    // Act
    const renderedTree = renderTree(buildShrinkTree(arb, value, { numItems: 100 })).join('\n');

    // Assert
    expect(arb.canShrinkWithoutContext(rawValue)).toBe(true);
    expect(renderedTree).toMatchSnapshot();
  });
});
