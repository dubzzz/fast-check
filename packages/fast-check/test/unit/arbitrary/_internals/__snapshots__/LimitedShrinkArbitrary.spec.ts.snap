// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LimitedShrinkArbitrary (integration) > should be able to shrink 251 given constraints maxShrinks:+0 1`] = `"251"`;

exports[`LimitedShrinkArbitrary (integration) > should be able to shrink 251 given constraints maxShrinks:1 1`] = `
"251
└> 0"
`;

exports[`LimitedShrinkArbitrary (integration) > should be able to shrink 251 given constraints maxShrinks:2 1`] = `
"251
├> 0
└> 126"
`;

exports[`LimitedShrinkArbitrary (integration) > should be able to shrink 251 given constraints maxShrinks:3 1`] = `
"251
├> 0
├> 126
|  └> 63
└> 189"
`;

exports[`LimitedShrinkArbitrary (integration) > should be able to shrink 251 given constraints maxShrinks:4 1`] = `
"251
├> 0
├> 126
|  ├> 63
|  |  └> 32
|  └> 95
├> 189
|  └> 158
└> 220"
`;
